pkg/virt-launcher/virtwrap/live-migration-source.go

```diff
@@ -530,103 +535,133 @@ func (l *LibvirtDomainManager) asyncMigrationAbort(vmi *v1.VirtualMachineInstanc
+		newProxy := migrationproxy.NewTargetProxy(loopbackAddress, port, nil, nil, migrationproxy.SourceUnixFile(l.virtShareDir, key))
+		proxies = append(proxies, newProxy)
+	}
 
-		domName := api.VMINamespaceKeyFunc(vmi)
-		dom, err := l.virConn.LookupDomainByName(domName)
-		if err != nil {
-			log.Log.Object(vmi).Reason(err).Error("Live migration failed.")
-			l.setMigrationResult(vmi, true, fmt.Sprintf("%v", err), "")
-			return
-		}
+	//  proxy incoming migration requests on port 22222 to the vmi's existing libvirt connection
+	libvirtConnectionProxy := migrationproxy.NewTargetProxy(loopbackAddress, LibvirtLocalConnectionPort, nil, nil, migrationproxy.SourceUnixFile(l.virtShareDir, string(vmi.UID)))
+	proxies = append(proxies, libvirtConnectionProxy)
+
+	return proxies
+}
+
+func generateMigrationParams(dom cli.VirDomain, vmi *v1.VirtualMachineInstance, options *cmdclient.MigrationOptions) (*libvirt.DomainMigrateParameters, error) {
+
+	migrURI := fmt.Sprintf("tcp://%s", ip.NormalizeIPAddress(ip.GetLoopbackAddress()))
+
+	bandwidth, err := converter.QuantityToMebiByte(options.Bandwidth)
+	if err != nil {
+		return nil, err
+	}
+
+	xmlstr, err := migratableDomXML(dom, vmi)
+	if err != nil {
+		return nil, err
+	}
+
+	params := &libvirt.DomainMigrateParameters{
+		Bandwidth:  bandwidth, // MiB/s
+		URI:        migrURI,
+		URISet:     true,
+		DestXML:    xmlstr,
+		DestXMLSet: true,
+	}
+
+	copyDisks := getDiskTargetsForMigration(dom, vmi)
+	if len(copyDisks) != 0 {
+		params.MigrateDisks = copyDisks
+		params.MigrateDisksSet = true
+	}
+
+	return params, nil
+}
+
+func (l *LibvirtDomainManager) asyncMigrate(vmi *v1.VirtualMachineInstance, options *cmdclient.MigrationOptions) error {
+
+	// get connection proxies for tunnelling migration through virt-handler
+	proxies := l.generateMigrationProxies(vmi)
+
+	migrateFlags := generateMigrationFlags(isBlockMigration(vmi), options.UnsafeMigration, options.AllowAutoConverge, options.AllowPostCopy)
+	if options.UnsafeMigration {
+		log.Log.Object(vmi).Info("UNSAFE_MIGRATION flag is set, libvirt's migration checks will be disabled!")
+	}
+
+	domName := api.VMINamespaceKeyFunc(vmi)
+	dom, err := l.virConn.LookupDomainByName(domName)
+	if err != nil {
+		return err
+	}
+
+	log.Log.Object(vmi).Infof("Initiating live migration.")
+	go func() {
 		defer dom.Free()
 
-		bandwidth, err := converter.QuantityToMebiByte(options.Bandwidth)
-		if err != nil {
-			log.Log.Object(vmi).Reason(err).Error("Live migration failed. Invalid bandwidth supplied.")
-			return
-		}
+		migrationErrorChan := make(chan error, 1)
+		defer close(migrationErrorChan)
 
-		if err := hotUnplugHostDevices(l.virConn, dom); err != nil {
-			log.Log.Object(vmi).Reason(err).Error(fmt.Sprintf("Live migration failed."))
-			l.setMigrationResult(vmi, true, fmt.Sprintf("%v", err), "")
-		}
+		// From here on out, any error encountered must be sent to the
+		// migrationError channel which is processed by the liveMigrationMonitor
+		// go routine.
```

Comment:
For the long run, assuring this instruction will be followed by authors and reviewers is hard.
How about we enforce this by a wrapping function.

Something like this:
```golang
withLiveMigrationMonitoring(func () error {
    dom := ...

    p, err := foo()
    if err != nil { return err }
    ...
    
})
```

Where `withLiveMigrationMonitoring` will create the channel and call the monitor go-routine and redirect any error from the inputted function to the channel:
```golang
func withLiveMigrationMonitoring(f)
   migrationErrorChan := make(chan error, 1)
   defer close(migrationErrorChan)

   go liveMigrationMonitor(vmi, l, options, migrationErrorChan)
   
   if err := f(); err !=nil {
       log...
       migrationErrorChan <- err
       return
   }
   ...
   
   WDYT?
```

Reply:
oh yeah. that's a pretty good idea. I'll do something like that. great suggestion

Reply:
fixed, this really helped simplify the code i think, thanks